### PR TITLE
impl: set useful gRPC keepalive defaults

### LINKED ISSUE #9842: Set helpful gRPC keepalive defaults
Use the defaults from #9832 across all libraries.

Here is an example of how we can set defaults, while respecting customer-supplied options:
https://github.com/googleapis/google-cloud-cpp/blob/f779bb99395bbdfe3aac51864ff2abf2a0b6dffc/google/cloud/bigtable/internal/defaults.cc#L134-L150

### PR DESCRIPTION
**Pull request by antonjfernando2021**
Fixes #9842 

notes: Hi, I am getting this exception, any help on this? 
```
46/409 Test  #46: common_grpc_utils_connection_options_test ........................................*****Exception: SegFault**  0.09 sec

The following tests FAILED:
	 46 - common_grpc_utils_connection_options_test (SEGFAULT)
```
notes: the build command and uni-test command respectively. 
```
~/google-cloud-cpp (keepalive_defaults_9842)$ cmake --build cmake-out/home
~/google-cloud-cpp (keepalive_defaults_9842)$ env -C cmake-out/home ctest --output-on-failure -LE integration-test
```

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10004)

### CODE REVIEW

File: google/cloud/connection_options_test.cc
Comment (antonjfernando2021) [inline google/cloud/connection_options_test.cc:169]: Attempted to DISABLE this unit-test, of course will remove _DISABLEd.

Comment (dbolduc) [reply google/cloud/connection_options_test.cc:169]: Looking...

I think this is an old test, written before we introduced the `Get*ChannelArgument()` helpers: https://github.com/googleapis/google-cloud-cpp/blob/dd4758319cdd6f3cffb5c339b07277075d79724f/google/cloud/grpc_options.h#L207-L213

I'm guessing that it segfaults because we are trying to access the wrong type (e.g. string vs. int) from this weird union thingy: https://grpc.github.io/grpc/cpp/uniongrpc__arg_1_1grpc__arg__value.html

We should use `GetStringChannelArguments()` instead of calling `actual.c_channel_args()`. I think the best thing to do is to refactor the two ChannelArguments tests. I would do this in a separate PR. It would be something like this:

```
auto actual = conn_opts.CreateChannelArguments();
auto user_agent = GetStringArgument(actual, "grpc.primary_user_agent"));
ASSERT_TRUE(user_agent.has_value());
EXPECT_THAT(*user_agent, StartsWith(conn_opts.user_agent_prefix()));
```

**Comment by dbolduc** [comment]
By the way, thanks for the PR!

I edited the title to pass the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) build. (we use `impl`, short for "implementation" to denote an internal feature that is not user-facing). I also linked the original issue in the description, such that it will be closed when this PR goes through.

I'll run the builds, just to do it. But we still need to update `connection_options_test.cc`, either in this PR or in a separate one.

**Comment by antonjfernando2021** [comment]
Fro the 2nd push: all unit-tests passed, Thanks for your suggestions on this:
        Start  46: common_grpc_utils_connection_options_test
 46/409 Test  #46: common_grpc_utils_connection_options_test ........................................   Passed    0.01 sec

100% tests passed, 0 tests failed out of 409

File: google/cloud/connection_options_test.cc
Comment (dbolduc) [inline google/cloud/connection_options_test.cc:190]: nit: I think this will be easier to read as:
```
EXPECT_EQ(*testing_pool, "testing-pool");
```

File: google/cloud/grpc_options.cc
Comment (dbolduc) [inline google/cloud/grpc_options.cc:46]: can we also copy over the comments, so we remember why we are using these settings.
https://github.com/googleapis/google-cloud-cpp/blob/315c752b080fc87350bf174697be8ace783efb08/google/cloud/storage/internal/storage_stub_factory.cc#L49-L59

Comment (dbolduc) [reply google/cloud/grpc_options.cc:46]: Also, can you remove that code block from `storage/internal/storage_stub_factory.cc`, now that the logic is handled in here?

File: google/cloud/grpc_options.cc
Comment (dbolduc) [inline google/cloud/grpc_options.cc:49]: I think the users should be able to override these values. (maybe they are super smart and know better than to use our defaults). So I think we should only set the defaults, if there is not a value set already. Something like this:

https://github.com/googleapis/google-cloud-cpp/blob/f779bb99395bbdfe3aac51864ff2abf2a0b6dffc/google/cloud/bigtable/internal/defaults.cc#L145-L150

 (the code in `storage` did not do this, but I think it should have).

File: google/cloud/grpc_options_test.cc
Comment (dbolduc) [inline google/cloud/grpc_options_test.cc:111]: A few comments here....

I think we should move these into a new `TEST(GrpcChannelArguments, MakeChannelArgumentsDefaults) {}`. It's just good to have focused tests. That way if one of the tests fail, we will know whether it is the defaulting logic or the combining logic.

I think the defaults test should verify that...
1. if a user does not set the values, our defaults are used.
2. if a user does set the values, their values are used. (not our defaults).

For part 1, I think we do not need to be checking for equality of the values. That is, our default keepalive time is 24 hours. Would we care if it was actually 23 hours? or 25 hours? I don't think so. I think it is sufficient to check that *any* value is set. Alternatively, we could check that the value is at least 1 hour or something.
```cc
EXPECT_GT(value, std::chrono::milliseconds(std::chrono::hours(1)).count());
```

**Review by dbolduc — CHANGES_REQUESTED** [review]
Thanks for cleaning up the `connection_options_test.cc`.

We are nearly there on the PR, just need to add defaulting logic.

**Comment by antonjfernando2021** [comment]
Thanks for the review comments Darren, hopefully we're almost there. This is my 1st open source contribution, I see lots of potential to learn here. But on Oct/11th starting new contract job, so my participation here might be limited.
Greetings from Ottawa/Canada.

File: google/cloud/grpc_options_test.cc
Comment (dbolduc) [inline google/cloud/grpc_options_test.cc:111]: Ahhhh... I think I gave you bad advice. I now see that these two lines are necessary, because we are checking for equality of the `grpc::ChannelArguments` objects. Let's call this the defaults test where no user option is supplied.

I have a nit though. If I stumbled on the code it would not be obvious what `86400000` means. Can we instead say: `std::chrono::milliseconds(std::chrono::hours(24)).count()`? and do the same for `60000`?

File: google/cloud/grpc_options_test.cc
Comment (dbolduc) [inline google/cloud/grpc_options_test.cc:139]: Let's set the channel arguments here (to something other than the defaults) e.g.
```cc
grpc::ChannelArguments args;
args.SetInt(GRPC_ARG_KEEPALIVE_TIME_MS, 1000);
args.SetInt(GRPC_ARG_KEEPALIVE_TIME_MS, 2000);
auto options = Options{}.set<GrpcChannelArgumentsNativeOption>(args);
```

then check that they are the same on the other side of `MakeChannelArguments`.

File: google/cloud/grpc_options_test.cc
Comment (dbolduc) [inline google/cloud/grpc_options_test.cc:138]: Let's rename this one to `MakeChannelArgumentsDefaults`

File: google/cloud/grpc_options_test.cc
Comment (dbolduc) [inline google/cloud/grpc_options_test.cc:142]: nit: I know I said `value` earlier, but maybe `auto keepalive_time` is a better name?

File: google/cloud/grpc_options_test.cc
Comment (dbolduc) [inline google/cloud/grpc_options_test.cc:146]: maybe: `auto keepalive_timeout` ?

**Review by dbolduc — CHANGES_REQUESTED** [review]
> But on Oct/11th starting new contract job

Congrats! I bet it pays better than open source contributing. 😆

> so my participation here might be limited

No worries, thanks for what you have done so far!

<hr />

Before we can merge the PR, you will need to sign the CLA (an agreement that lets Google use your code) as described in https://github.com/googleapis/google-cloud-cpp/pull/10004#issuecomment-1272180470.

The other thing is that the `checkers` build is failing because of unrelated changes to this PR. It is fixed at HEAD, but you will need to rebase to pick up those changes. You can do that with:
```sh
git fetch upstream
git rebase upstream/main
```

**Comment by antonjfernando2021** [comment]
Hi Darren, so I submitted the individual contributor CLA, and rebased upstream etc, and added new review suggestions. When native settings are set, I was expecting 1000/2000 respectively, not the 24hrs/60secs, so something not right, I am saving the work for now, may be look at next weekend.  The default 24hrs/60secs are set only if nothing else exist already. Thanks.

**Review by dbolduc — CHANGES_REQUESTED** [review]
Thanks for signing the CLA. We are so close, I swear. I will approve if you

- [ ] rename the one test
- [ ] delete the 3 other tests
- [ ] then run the linter

File: google/cloud/grpc_options_test.cc
Comment (antonjfernando2021) [reply google/cloud/grpc_options_test.cc:161]: Hi, after rebasing, seen some compile issues, not related to my changes. 

-VirtualBox ~/google-cloud-cpp (keepalive_defaults_9842)$ git remote -v
origin	git@github.com:antonjfernando2021/google-cloud-cpp.git (fetch)
origin	git@github.com:antonjfernando2021/google-cloud-cpp.git (push)
upstream	git@github.com:googleapis/google-cloud-cpp.git (fetch)
upstream	git@github.com:googleapis/google-cloud-cpp.git (push)
-VirtualBox ~/google-cloud-cpp (keepalive_defaults_9842)$ git fetch upstream 
-VirtualBox ~/google-cloud-cpp (keepalive_defaults_9842)$ git rebase upstream/main
Current branch keepalive_defaults_9842 is up to date.

COMPILE: 

[ 60%] Building CXX object google/cloud/iam/CMakeFiles/google_cloud_cpp_iam.dir/iam_client.cc.o
In file included from /home/anton/google-cloud-cpp/google/cloud/iam/iam_connection.h:22,
                 from /home/anton/google-cloud-cpp/google/cloud/iam/iam_client.h:22,
                 from /home/anton/google-cloud-cpp/google/cloud/iam/iam_client.cc:19:
/home/anton/google-cloud-cpp/google/cloud/iam/iam_connection_idempotency_policy.h:80:31: error: ‘google::iam::admin::v1::DisableServiceAccountKeyRequest’ has not been declared
   80 |       google::iam::admin::v1::DisableServiceAccountKeyRequest const& request);
      |                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/anton/google-cloud-cpp/google/cloud/iam/iam_connection_idempotency_policy.h:83:31: error: ‘google::iam::admin::v1::EnableServiceAccountKeyRequest’ has not been declared
   83 |       google::iam::admin::v1::EnableServiceAccountKeyRequest const& request);
      |                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/anton/google-cloud-cpp/google/cloud/iam/iam_connection.h:24,
                 from /home/anton/google-cloud-cpp/google/cloud/iam/iam_client.h:22,
                 from /home/anton/google-cloud-cpp/google/cloud/iam/iam_client.cc:19:
/home/anton/google-cloud-cpp/google/cloud/iam/internal/iam_stub.h:99:31: error: ‘google::iam::admin::v1::DisableServiceAccountKeyRequest’ has not been declared
   99 |       google::iam::admin::v1::DisableServiceAccountKeyRequest const&
      |                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/anton/google-cloud-cpp/google/cloud/iam/internal/iam_stub.h:104:31: error: ‘google::iam::admin::v1::EnableServiceAccountKeyRequest’ has not been declared
  104 |       google::iam::admin::v1::EnableServiceAccountKeyRequest const&

Comment (coryan) [reply google/cloud/grpc_options_test.cc:161]: I think the CMake cache is too sticky, and you started working before we fixed that problem.

Try removing `CMakeCache.txt` in  $CMAKE_BINARY_DIR .  Or just remove the whole build directory if you have to.

Comment (antonjfernando2021) [reply google/cloud/grpc_options_test.cc:161]: cool thanks Darren, progressing but now out of disk space, had only 100G for vm. Once I figured how to increase vm disk space, will get back to this, likely next weekend, luckily laptop still got about 281G left. 
[ 57%] Building CXX object google/cloud/bigquery/CMakeFiles/google_cloud_cpp_bigquery.dir/internal/bigquery_write_logging_decorator.cc.o
fatal error: error writing to /tmp/ccyEqzlu.s: No space left on device)

Comment (antonjfernando2021) [reply google/cloud/grpc_options_test.cc:161]: removed a comment.

**Comment by antonjfernando2021** [comment]
Hi, only running the formatter failed, rest of the suggestions fixed. Thanks. 

-VirtualBox ~/google-cloud-cpp (keepalive_defaults_9842)$ **ci/cloudbuild/build.sh -t checkers-pr**

2022-10-10T23:40:16Z (+0s)
=======================================
|   Starting docker build: checkers   |
=======================================

2022-10-10T23:40:16Z (+0s)
--------------------------------------------------
|   Building docker image: gcb-checkers:latest   |
--------------------------------------------------
2022-10-10T23:40:16Z (+0s): + docker build -t gcb-checkers:latest --build-arg=NCPU=4 -f ci/cloudbuild/dockerfiles/checkers.Dockerfile ci
/home/anton/google-cloud-cpp/ci/lib/io.sh: line 110: docker: command not found

**Review by dbolduc — APPROVED** [review]
:lgtm:

Sorry that the formatting effort was so annoying. I just ran the tool myself and added a commit onto your branch.

Thanks for fixing this issue! And good luck at the new job!

> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10004#-:-NE4HlUo3VUl9ViG464u:byr3xe0)* status: 0 of 4 files reviewed, 16 unresolved discussions (waiting on @antonjfernando2021 and @coryan)

**Comment by antonjfernando2021** [comment]
> Thanks for fixing this issue! And good luck at the new job!

Thanks Darren.

**Pull request merged**
